backend/src/apiserver/archive/log.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package archive
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+)
+
+type LogFormat string
+
+const (
+	LogFormatJSON = LogFormat("json-lines")
+	LogFormatText = LogFormat("text")
+)
+
+type ExtractLogOptions struct {
+	LogFormat  LogFormat
+	Timestamps bool
+}
+
+type LogArchiveInterface interface {
+	GetLogObjectKey(workflow util.Workflow, nodeId string) (string, error)
+	CopyLogFromArchive(logContent []byte, dst io.Writer, opts ExtractLogOptions) error
+}
+
+// Log Archive
+type RunLogEntry struct {
+	Log       string    `json:"log"`
+	Timestamp time.Time `json:"timestamp,omitempty"`
+}
+
+var k8sLogPrefixExp = regexp.MustCompile(`(?m)^(\d{4}-\d{2}-\d{2}T\S+)\s(.+)$`)
+var crioLogPrefixExp = regexp.MustCompile(`(?m)^(.+)\s(stdout|stderr)\s\w\s(.+)$`)
+
+type LogArchive struct {
+}
+
+func NewLogArchive() *LogArchive {
+	return new(LogArchive)
+}
+
+func (a *LogArchive) GetLogObjectKey(workflow util.Workflow, nodeId string) (key string, err error) {
+	taskRun, taskRunId := workflow.FindTaskRunByPodName(nodeId)
+	if taskRun == nil {
+		err = util.NewBadRequestError(errors.New("archived log cannot be read"),
+			"Failed to retrieve the task run from the runtime workflow")
+		return
+	}
+
+	workflowName := workflow.Name
+	taskName := taskRun.PipelineTaskName
+
+	key = fmt.Sprintf("/artifacts/%s/%s/%s--step-main_log.gz", workflowName, taskName, taskRunId)
+
+	return
+}
+
+// CopyLogFromArchive copies a task run archived log into expected format.
+func (a *LogArchive) CopyLogFromArchive(logContent []byte, dst io.Writer, opts ExtractLogOptions) error {
+	reader, err := decompressLogArchive(logContent)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		// line := strings.Trim(scanner.LogFormatText(), "\n\r\t ")
```

Comment:
Is this commented code for debugging?

Reply:
Thanks, that's redundant comment. I've already removed it.